hash_practice/exercises.py

```diff
@@ -2,18 +2,43 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O (n^2(logn))because I'm using the sorted function inside the loop 
+        Space Complexity: O(n) because I'm creating a dictionary and looking over the values
     """
-    pass
+    anagrams={}
+    anagram_list=[]
+    for word in strings:
+        sorted_word="".join(sorted(word))
+        if anagrams.get(sorted_word):
+            anagrams[sorted_word].append(word)
+        else:
+            anagrams[sorted_word]=[word]
+    for word in anagrams.values():
+        anagram_list.append(word)
+    return anagram_list
+
+        
+    
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n log n) because I'm using the sorted function to sort the dictionary
+        Space Complexity: O(log n) because I'm storing the results in a list that loops through the sorted dictionary which is compacted already

```

Comment:
I'm not certain you can consider a dictionary "compacted". Not that we expect you to be fully aware but in general, dictionaries still take O(n) space, as all the entries are somewhere in the dictionary, there's no "compacting". A dictionary can usually achieve O(1) look-up and insertion time not because of "compaction" but a hash function applied to the keys is used to add and retrieve entries. (Yes, in fact dictionaries are hash tables as you learned in this lesson!)

For a solution to have O(log(n)) space complexity seems unlikely because that means you would have to store only a small subset of items but still also have all the items accessible. In general, you only see O(log(n)) space complexity when analyzing recursive functions (for example functions working with binary search trees), as then only O(log(n)) items may be placed on the stack as the recursion is performed.

In this case, I would say the space complexity is still O(n). In the worst case for space, every entry in `nums` appears once, so `freq_dict` will have O(n) key-value pairs.